### PR TITLE
feat: monograph support

### PR DESCRIPTION
**Pull request by thisisnithin**
## Motivation and Context

image

image

## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: studio/src/components/federatedgraphs-cards.tsx [hunk 1]
```diff
@@ -463,25 +463,39 @@ export const Empty = ({
...
+            docs
+          </a>{" "}
+          ).
         </>
       }
       actions={
         <div className="flex flex-col gap-y-6">
           <CLI
-            command={`npx wgc federated-graph create production --namespace ${router.query.namespace} --label-matcher ${labels} --routing-url http://localhost:4000/graphql`}
+            command={`npx wgc federated-graph create production --namespace ${router.query.namespace} --label-matcher ${labels} --routing-url http://localhost:3002/graphql`}
```
Comment (StarpTech) [inline studio/src/components/federatedgraphs-cards.tsx:495]: We should show a dropdown from where the user can select the command for the graph type. Showing two long cli commands can be overwhelming.

File: studio/src/components/layout/graph-layout.tsx [hunk 2]
```diff
@@ -227,6 +231,18 @@ export const GraphSelect = () => {
       value={selected?.id}
       onValueChange={(gID) => {
         const graph = data?.graphs.find((g) => g.id === gID);
+
+        // don't show subgraphs and compositions page for monographs
+        if (
+          graph?.type === "graph" &&
```
Comment (StarpTech) [inline studio/src/components/layout/graph-layout.tsx:237]: We should rename `graph` to `monograph` to make it very explicit.

Comment (StarpTech) [reply studio/src/components/layout/graph-layout.tsx:237]: Composition page is still valuable for monographs. Also the recent added feature "Config Signing & Validation" requires it which works with both graph types.

File: studio/src/pages/[organizationSlug]/[namespace]/graph/[slug]/changelog/index.tsx [hunk 3]
```diff
@@ -320,7 +320,11 @@ const ChangelogPage: NextPageWithLayout = () => {
           }
           actions={
             <CLI
-              command={`npx wgc subgraph publish users --namespace ${router.query.namespace} --schema users.graphql`}
+              command={
+                graphData?.graph?.type === "federated"
+                  ? `npx wgc subgraph publish <subgraph-name> --namespace ${router.query.namespace} --schema <path-to-schema>`
```
Comment (StarpTech) [inline studio/src/pages/[organizationSlug]/[namespace]/graph/[slug]/changelog/index.tsx:325]: Same here. Provide a dropdown.

**Review by StarpTech — CHANGES_REQUESTED** [review]
Please don'r forget to add integration tests for the new monographs capabilities.

File: cli/src/commands/graph/monograph/commands/publish.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    const schemaBuffer = await readFile(schemaFile);
+    const schema = new TextDecoder().decode(schemaBuffer);
+    if (schema.trim().length === 0) {
+      program.error(
+        pc.red(pc.bold(`The schema file '${pc.bold(schemaFile)}' is empty. Please provide a valid schema.`)),
+      );
+    }
+
+    const spinner = ora('Subgraph is being published...').start();
```
Comment (StarpTech) [inline cli/src/commands/graph/monograph/commands/publish.ts:49]: Typo

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: studio/src/pages/[organizationSlug]/[namespace]/graph/[slug]/changelog/index.tsx (hunk 3 above)
Comment (thisisnithin) [reply studio/src/pages/[organizationSlug]/[namespace]/graph/[slug]/changelog/index.tsx:325]: This one shouldn't be a dropdown. We only show one of them.

File: cli/src/commands/graph/monograph/commands/migrate.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      },
+      {
+        headers: baseHeaders,
+      },
+    );
+
+    spinner.stop();
+
+    if (resp.response?.code === EnumStatusCode.OK) {
+      console.log(
```
Comment (StarpTech) [inline cli/src/commands/graph/monograph/commands/migrate.ts:39]: Use spinner.succes und fail

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: studio/src/pages/[organizationSlug]/[namespace]/graph/[slug]/changelog/index.tsx (hunk 3 above)
Comment (StarpTech) [reply studio/src/pages/[organizationSlug]/[namespace]/graph/[slug]/changelog/index.tsx:325]: ok

File: controlplane/src/db/models.ts [hunk 6]
```diff
@@ -38,7 +39,7 @@ export type AuditableType =
   | 'operation_change_override'
   | 'operation_ignore_all_override';
 
-export type AuditTargetType = 'organization' | 'subgraph' | 'federated_graph' | 'user';
+export type AuditTargetType = 'organization' | 'subgraph' | 'federated_graph' | 'monograph' | 'user';
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/db/models.ts:42]: Target id for audit logs will always be federated graph id. The type is shown as is on the frontend which is why we have monograph.

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 7]
```diff
@@ -7383,7 +8148,9 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
 
         const federatedGraphs = await fedRepo.getAccessibleFederatedGraphs(authContext.userId);
 
-        const subgraphs = await subgraphRepo.getAccessibleSubgraphs(authContext.userId);
+        const subgraphs = (await subgraphRepo.getAccessibleSubgraphs(authContext.userId)).filter(
+          (s) => s.labels.length > 0 && s.labels[0].key !== 'monograph',
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:8152]: I did not understand, could you please explain

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 8]
```diff
@@ -1736,18 +2251,21 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
-            federated_graph: {
-              id: graph.id,
-              name: graph.name,
-              namespace: graph.namespace,
-            },
-            organization: {
-              id: authContext.organizationId,
-              slug: authContext.organizationSlug,
+          orgWebhooks.send({
+            eventName: OrganizationEventName.FEDERATED_GRAPH_SCHEMA_UPDATED,
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:2255]: It is correct, maybe wrapping function name not clear in the diff view

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 9]
```diff
@@ -2616,18 +3365,21 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
-          federated_graph: {
-            id: federatedGraph.id,
-            name: federatedGraph.name,
-            namespace: federatedGraph.namespace,
-          },
-          organization: {
-            id: authContext.organizationId,
-            slug: authContext.organizationSlug,
+        orgWebhooks.send({
+          eventName: OrganizationEventName.FEDERATED_GRAPH_SCHEMA_UPDATED,
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:3369]: Same as above

File: cli/src/commands/graph/monograph/commands/check.ts [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    const lintIssuesTable = new Table({
+      head: [pc.bold(pc.white('LINT_RULE')), pc.bold(pc.white('ERROR_MESSAGE')), pc.bold(pc.white('LINE NUMBER'))],
+      colAligns: ['left', 'left', 'center'],
+      wordWrap: true,
+    });
+
+    let success = false;
+    let finalStatement = '';
+
+    let studioCheckDestination = '';
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply cli/src/commands/graph/monograph/commands/check.ts]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 11]
```diff
@@ -5494,17 +6252,19 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
-            name: g.name,
-            routingURL: g.routingUrl,
-            lastUpdatedAt: g.lastUpdatedAt,
-            labels: g.labels,
-            createdUserId: g.creatorUserId,
-            targetId: g.targetId,
-            subscriptionUrl: g.subscriptionUrl,
-            namespace: g.namespace,
-          })),
+          graphs: list
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 12]
```diff
@@ -7359,10 +8122,12 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
-          const subgraphs = await subgraphRepo.list({
-            limit: 0,
-            offset: 0,
-          });
+          const subgraphs = (
+            await subgraphRepo.list({
+              limit: 0,
+              offset: 0,
+            })
+          ).filter((s) => s.labels.length > 0 && s.labels[0].key !== 'monograph');
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 13]
```diff
@@ -7383,7 +8148,9 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
 
         const federatedGraphs = await fedRepo.getAccessibleFederatedGraphs(authContext.userId);
 
-        const subgraphs = await subgraphRepo.getAccessibleSubgraphs(authContext.userId);
+        const subgraphs = (await subgraphRepo.getAccessibleSubgraphs(authContext.userId)).filter(
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Done

File: controlplane/src/core/repositories/FederatedGraphRepository.ts [hunk 14]
```diff
@@ -294,6 +297,23 @@ export class FederatedGraphRepository {
       .where(and(eq(targets.id, targetId), eq(schema.targets.organizationId, this.organizationId)));
   }
 
+  // We do not need to do any recomposition since we move both the graph and subgraph to the new namespace.
+  // The namespaceId column update is enough.
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/repositories/FederatedGraphRepository.ts:303]: Done

File: controlplane/src/core/util.ts [hunk 15]
```diff
@@ -97,6 +97,8 @@ export function normalizeLabels(labels: Label[]): Label[] {
   return [...uniqueLabels].map((label) => splitLabel(label));
 }
 
+const reservedKeys = new Set(['monograph']);
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/util.ts]: Done

File: studio/src/components/federatedgraphs-cards.tsx (hunk 1 above)
Comment (thisisnithin) [reply studio/src/components/federatedgraphs-cards.tsx:495]: Done

File: studio/src/components/layout/graph-layout.tsx (hunk 2 above)
Comment (thisisnithin) [reply studio/src/components/layout/graph-layout.tsx]: Thanks, I did make the page visible but forgot about this redirect. Done.

File: cli/src/commands/graph/monograph/commands/publish.ts (hunk 4 above)
Comment (thisisnithin) [reply cli/src/commands/graph/monograph/commands/publish.ts]: Fixed

File: cli/src/commands/graph/monograph/commands/migrate.ts (hunk 5 above)
Comment (thisisnithin) [reply cli/src/commands/graph/monograph/commands/migrate.ts]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 16]
```diff
@@ -5569,6 +6336,8 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
           limit: req.limit,
           offset: req.offset,
           namespaceId: namespace?.id,
+          excludeMonographs: req.monographConfig.case === 'excludeMonographs' && req.monographConfig.value,
```
Comment (StarpTech) [inline controlplane/src/core/bufservices/PlatformService.ts:6339]: Also here. Now a getFederatedGraphs method has to deal with monographs too. We all know why but this vague line is hard to reason about. Can we rename `getFederatedGraphs` to `getGraphs`? Alternatively, introduce new RPC methods to deliver only monographs and federated graphs.

File: controlplane/src/core/repositories/FederatedGraphRepository.ts [hunk 17]
```diff
@@ -294,6 +298,24 @@ export class FederatedGraphRepository {
       .where(and(eq(targets.id, targetId), eq(schema.targets.organizationId, this.organizationId)));
   }
 
+  // We do not need to do any recomposition since we move both the graph and subgraph to the new namespace.
+  // Also, due to unique labels, they will not compose with other graphs.
+  // The namespaceId column update is enough.
+  public async moveMonograph({
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FederatedGraphRepository.ts:304]: As described earlier, we need to find a better abstraction in the code base to not shot our self in the foot. Monographs are federated graphs with just one subgraph but users will work with them as a monograph.  We're trying to merge them somehow inn the codebase without clear boundaries. There are some differences when dealing with both types and this needs to be respected. I'd recommend to separate them on the service and repository level where possible. You can put these new methods in a `MonographRepository`

File: controlplane/src/core/repositories/FederatedGraphRepository.ts [hunk 18]
```diff
@@ -522,7 +584,7 @@ export class FederatedGraphRepository {
       .where(
         and(
           eq(targets.organizationId, this.organizationId),
-          eq(targets.type, 'federated'),
+          eq(schema.targets.type, 'federated'),
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FederatedGraphRepository.ts:587]: Revert? Targets must work.

File: controlplane/src/core/repositories/SubgraphRepository.ts [hunk 19]
```diff
@@ -95,6 +96,7 @@ export class SubgraphRepository {
           organizationId: this.organizationId,
           labels: uniqueLabels.map((ul) => joinLabel(ul)),
           readme: data.readme,
+          asMonograph: data.asMonograph,
```
Comment (StarpTech) [inline controlplane/src/core/repositories/SubgraphRepository.ts:99]: This is another abstraction I find odd. A subgraph belongs to a federated graph but does not store metadata about the graph it belongs to. For me it feels like, that this information should not be needed, and the right subgraph is joined when selecting a target over the type `monograph`

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: cli/src/commands/graph/monograph/commands/delete.ts [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+        headers: baseHeaders,
+      },
+    );
+
+    if (resp.response?.code === EnumStatusCode.OK) {
+      spinner.succeed(`Monograph was deleted successfully.`);
+    } else {
+      spinner.fail(`Failed to delete monograph.`);
+      if (resp.response?.details) {
+        console.log(pc.red(pc.bold(resp.response?.details)));
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply cli/src/commands/graph/monograph/commands/delete.ts:44]: Done

File: controlplane/migrations/0084_goofy_raider.sql [hunk 21]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE "targets" ALTER COLUMN "type" SET NOT NULL;--> statement-breakpoint
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/migrations/0084_goofy_raider.sql:1]: Cleaned it up

File: controlplane/test/monograph.test.ts [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+    expect(monographRes.graph?.labelMatchers[0]).toContain('_internal=');
+    expect(monographRes.graph?.routingURL).toBe('http://localhost:3002');
+    expect(monographRes.subgraphs.length).toBe(1);
+    expect(monographRes.subgraphs[0].labels.length).toBe(1);
+    expect(monographRes.subgraphs[0].labels[0].key).toBe('_internal');
+    expect(monographRes.subgraphs[0].routingURL).toBe('http://localhost:4000');
+
+    await server.close();
+  });
+
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/test/monograph.test.ts:47]: Added the test in `check-subgraph-schema` test file

File: controlplane/src/core/repositories/FederatedGraphRepository.ts [hunk 23]
```diff
@@ -498,12 +513,21 @@ export class FederatedGraphRepository {
     return graphs.length === 1;
   }
 
-  public byName(name: string, namespace: string): Promise<FederatedGraphDTO | undefined> {
+  public byName(
+    name: string,
+    namespace: string,
+    opts?: {
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/repositories/FederatedGraphRepository.ts:519]: Discussed

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**